sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusListener.cs

```diff
@@ -550,6 +591,262 @@ private async Task RunBatchReceiveLoopAsync(CancellationToken cancellationToken)
+                    }
+
+                    await Task.WhenAll(abandonTasks).ConfigureAwait(false);
+                }
+            }
+        }
+
+        private async Task MonitorCache(CancellationToken backgroundCancellationToken)
+        {
+            var acquiredSemaphore = false;
+            try
+            {
+                _cycleDuration = ValueStopwatch.StartNew();
+
+                // Wait max wait time after starting this task before checking the number of messages.
+                while (_cycleDuration.GetElapsedTime() < _serviceBusOptions.MaxWaitTime && !backgroundCancellationToken.IsCancellationRequested)
+                {
+                    var remainingTime = GetRemainingTime(_cycleDuration.GetElapsedTime());
+                    await Task.Delay(remainingTime, backgroundCancellationToken).ConfigureAwait(false);
+                }
+
+                // This will throw if the cancellation token has been canceled.
+                if (!_cachedMessagesGuard.Wait(0, backgroundCancellationToken))
+                {
+                    await _cachedMessagesGuard.WaitAsync(backgroundCancellationToken).ConfigureAwait(false);
+                }
+                acquiredSemaphore = true;
+
+                // Since max wait time has passed, pull all messages out of the cache and invoke the function on it.
+                var triggerMessages = _cachedMessagesManager.TryGetBatchofMessagesWithCached(allowPartialBatch: true);
+
+                if (triggerMessages.Length > 0)
+                {
+                    ServiceBusTriggerInput input = ServiceBusTriggerInput.CreateBatch(
+                            triggerMessages,
+                            _monitoringCycleMessageActions,
+                            _monitoringCycleReceiveActions,
+                            _client.Value);
+
+                    await TriggerWithMessagesInternal(triggerMessages, input, _monitoringCycleReceiver, _monitoringCycleReceiveActions, _monitoringCycleMessageActions, backgroundCancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    _logger.LogDebug($"Service Bus Listener has waited MaxWaitTime since last invocation but there are no messages still being held.");
+                }
+
+                // After one wait cycle, cancel the background task cancellation token source. It can be assumed that there will never be more than the
+                // minimum batch size number of messages in the cache. The monitoring cycle will be restarted by the receive batch loop if needed.
+                if (_backgroundCacheMonitoringCts != null)
+                {
+                    _backgroundCacheMonitoringCts.Cancel();
+                    _backgroundCacheMonitoringCts.Dispose();
+                    _backgroundCacheMonitoringCts = null;
+                }
+            }
+            catch (TaskCanceledException)
```

Comment:
This is redundant with the OperationCanceledException block as TaskCanceledException derives from it.